### ISSUE TITLE
Error graphiqlHapi: custom plugins assigned to graphiql route will be ignored

### ISSUE THREAD

**Issue by danielglennross** [apollographql/graphql-server#417]
When registering `graphiqlHapi`, if plugins are assigned to the graphiql options route - such as:

```
{
  register: graphiqlHapi,
  options: {
    path: '/graphiql',
    graphiqlOptions: {
      endpointURL: '/graphql'
    },
    route: {
      plugins: {
        myCustomPlugin: {}
      }
    }
  }
}
```

they will be overwritten with the `graphiql` plugin rather than being merged when the route is configured.

The following line seems to be the source of the issue:
https://github.com/apollographql/graphql-server/blob/master/packages/graphql-server-hapi/src/hapiApollo.ts#L83

**Comment by helfer** [comment]
Hi @danielglennross I'm not very familiar with hapi. Could you make an example of what should happen, and tell me what happens instead? If other people concur, it would be great if you could make a PR to fix the issue! 🙂

**Comment by danielglennross** [comment]
Looks like the situation will occur for both the `plugin` and `pre` route config options.
If a `graphiqlHapi` Hapi plugin is configured like:
```
{
  register: graphiqlHapi,
  options: {
    path: '/graphiql',
    graphiqlOptions: {
      endpointURL: '/graphql'
    },
    route: {
      plugins: {
        myCustomPlugin: {}
      },
      pre: [{
        assign: 'foobar'
        method: 'foobar'
      }]
    }
  }
}
```
These properties will be overwritten by the `plugin` and `pre` properties specified in the `Object.assign`, rather than merged with them.
So we get a Hapi route which ultimately looks like:
```
{
  method: 'GET',
  path: options.path || '/graphql',
  config: {
    plugins: {
      graphiql: options.graphiqlOptions
    },
    pre: [{
      assign: 'graphiqlParams',
      method: 'getGraphiQLParams',
    }, {
      assign: 'graphiQLString',
      method: 'renderGraphiQL(route, pre.graphiqlParams)',
    }]
  }
  handler: (request, reply) => {
    reply(request.pre['graphiQLString']).header('Content-Type', 'text/html');
  },
}
```
instead of:
```
{
  method: 'GET',
  path: options.path || '/graphql',
  config: {
    plugins: {
      graphiql: options.graphiqlOptions,
      myCustomPlugin: {}
    },
    pre: [{
      assign: 'graphiqlParams',
      method: 'getGraphiQLParams',
    }, {
      assign: 'graphiQLString',
      method: 'renderGraphiQL(route, pre.graphiqlParams)',
    }, {
      assign: 'foobar',
      method: 'foobar'
    }]
  }
  handler: (request, reply) => {
    reply(request.pre['graphiQLString']).header('Content-Type', 'text/html');
  },
}
```

**Comment by helfer** [comment]
Thanks for clarifying @danielglennross. Yes please, it would be great if you could make a PR to fix it! 🙂

**Comment by danielglennross** [comment]
@helfer looks like the open PR that allows GraphiQLOptions to be passed as a function will also resolve this issue: https://github.com/apollographql/graphql-server/pull/426/files#diff-15d7de851dd7ea8c03559922a37d4904R95
Would you prefer just to merge that?

**Comment by helfer** [comment]
@danielglennross sure, I'm hoping to merge that as soon as it's ready. Once we've merged it, we can hopefully close this issue! 🙂

**Comment by danielglennross** [comment]
@helfer can confirm this issue is resolved in `graphql-server-hapi@0.9.0`
